### PR TITLE
arch: riscv: reset global pointer on exception

### LINKED ISSUE #81372: arch: riscv: userspace: potential security risk when `CONFIG_RISCV_GP=y`
**Describe the bug**

When the Global Pointer (GP) relative addressing is enabled (`CONFIG_RISCV_GP=y`), the `gp` reg points at 0x800 bytes past the start of the `.sdata` section which is then used by the linker to relax accesses to global symbols.

https://github.com/zephyrproject-rtos/zephyr/blob/c0a0e6a3887c53e5519ae5b30352b39cf7e9288f/include/zephyr/arch/riscv/common/linker.ld#L305-L319

However, the `gp` reg is not protected against write from userspace, this means that a rogue userspace can corrupt the `gp` reg, and cause the compiled instruction to access random addresses.

**To Reproduce**
1. Build the `qemu_riscv64` board with userspace and `CONFIG_RISCV_GP` enabled
2. Note the value of the `gp` register
3. Write some random value to the `gp` reg from one userspace thread
4. Notice that the `gp` reg is now changed

**Expected behavior**

The `gp` register should remain a constant.

**Impact**

A rogue thread can corrupt the `gp` reg and cause the entire system to hard fault at best, at worst, it can potentially trick the system to access another set of random global symbols.

**Environment (please complete the following information):**

 - Toolchain (e.g Zephyr SDK, ...): 0.16.8
 - Commit SHA or Version used: v3.7 branch

### PR DESCRIPTION
**Pull request by ycsin**
Reset the gp on exception entry from u-mode to protect the kernel against a possible rogue user thread when global pointer (GP) relative addressing is enabled (`CONFIG_RISCV_GP=y`).

Fixes #81372

### CODE REVIEW

**Comment by ycsin** [comment]
cc @cwshu

**Review by npitre — DISMISSED** [review]

**Comment by ycsin** [comment]
Added test

File: arch/riscv/core/isr.S [hunk 1]
```diff
@@ -163,6 +163,14 @@ SECTION_FUNC(exception.entry, _isr_wrapper)
 	lr t0, ___cpu_t_current_OFFSET(s0)
 	lr tp, _thread_offset_to_tls(t0)
 
+	/* Make sure global pointer is sane */
+#ifdef CONFIG_RISCV_GP
+	.option push
+	.option norelax
+	la gp, __global_pointer$
```
Comment (cfriedt) [inline arch/riscv/core/isr.S:170]: Just some notes

It looks like GP is specified as neither callee saved nor caller saved in the riscv spec.

OpenSBI backs it up and restores GP in their handler.

https://www.openeuler.org/en/blog/zhangxinhao/risc-v/2020-11-28-RISC-V.html

I think we're under no obligation to restore it or back it up.

Some bare metal environments might use it.

https://five-embeddev.com/quickref/global_pointer.html


Is the point here to illustrate that we _can_ use it from within uninterruptible context?

**Comment by npitre** [comment]
Chris Friedt wrote:

> Is the point here to illustrate that we _can_ use it from within uninterruptible context?

If `CONFIG_RISCV_GP` is set then gp is used by all the code compiled by gcc to reference global variables.
I let you imagine how futile the security model is if user threads may mess with it.

File: arch/riscv/core/isr.S (hunk 1 above)
Comment (fkokosinski) [reply arch/riscv/core/isr.S:170]: GCC seems to never save GP: https://github.com/gcc-mirror/gcc/blob/master/gcc/config/riscv/riscv.cc#L7368-L7370

**Review by npitre — APPROVED** [review]
And yes, this should definitely be tagged as a security fix.

File: arch/riscv/core/isr.S (hunk 1 above)
Comment (cfriedt) [reply arch/riscv/core/isr.S:170]: It's probably smart to support at least one convention for consistency in the kernel.

I can also see kernel-side code potentially using  GP as well, but if the preempted threads value is destroyed by an ISR with higher priority, I would certainly want to have it back the way I expected.

Maybe this below? Is this what we already do?

* back up the old GP, so the preempted context can still have a consistent register view or return
* do what we do here, and set it to a default value on entering context (ISR can adjust however)
* restore on return from interrupt

@npitre, let me know if this aligns with your suggestion as well. I understand it's adding one more register to push / pop, but it's maybe worth it?

**Comment by cfriedt** [comment]
Just curious if we back up / restore GP already? If not, we may want to consider doing so.

**Comment by npitre** [comment]
On Tue, 12 Nov 2024, Chris Friedt wrote:

> @npitre, let me know if this aligns with your suggestion as well. I 
> understand it's adding one more register to push / pop, but it's maybe 
> worth it?

No it is not.

The gp, when so instructed by the build system, is relied upon by the 
linker to produce tighter code for addressing global variables. In such 
case it stands for "Global Pointer" and not "General Purpose". It is 
then used potentially everywhere a relocation is involved: isr, thread 
context, user context, etc. It is assumed by the linker to hold the same 
value everywhere, be set once and never modified afterwards. You cannot 
just accidentally "corrupt it (you're likely to crash right away). Even 
the assembler needs this norelax mode to touch it otherwise it complains 
with an error.

Please see 
https://www.sifive.com/blog/all-aboard-part-3-linker-relaxation-in-riscv-toolchain 
for more details.

So _preserving_ the gp is completely useless and wasteful. Restoring it 
is also wasteful _unless_ untrusted code may be involved. That untrusted 
code could potentially create a privilege elevation attack with a 
carefully chosen gp value and syscall combination to trick the 
privileged kernel code into updating the wrong global variable. That's 
the reason why exiting user mode must re-set the gp value.

Now if we had an hypervisor switching between different OSes then 
obviously that hypervisor would have to save and restore the gp value as 
it is going to differ between different systems. But within a given 
environment it is not meant to change.

**Review by fkokosinski — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Comment by ceolin** [comment]
> And yes, this should definitely be tagged as a security fix.

Agree ! this should be merged ASAP imho.

**Comment by bonetor** [comment]
Why is there no backport to v3.6 for the patch provided?

**Comment by ycsin** [comment]
> Why is there no backport to v3.6 for the patch provided?

3.6 isn't my focus so I probably forgot about it last time - I'd suggest anyone still on v3.6 to use v3.7 as the latter is LTS

**Pull request merged**